File: 3609.minimum-moves-to-reach-target-in-grid.go
#
# @lc app=leetcode id=3609 lang=golang
#
# [3609] Minimum Moves to Reach Target in Grid
#

# @lc code=start
func minMoves(sx int, sy int, tx int, ty int) int {
	if sx == tx && sy == ty {
		return 0
	}

	mod := int64(1000000007)

	keyFunc := func(x, y int) int64 {
		return int64(x)*mod + int64(y)
	}

	visited := make(map[int64]bool)
	queue := [][]int{{tx, ty, 0}}

	visited[keyFunc(tx, ty)] = true

	getPrevs := func(x, y int) [][]int {
		res := [][]int{}
		// From x direction halve
		if x%2 == 0 {
			m := x / 2
			if m >= y {
				res = append(res, []int{m, y})
			}
		}
		// From x direction subtract
		if x >= y && x <= 2*y {
			res = append(res, []int{x - y, y})
		}
		// From y direction halve
		if y%2 == 0 {
			m := y / 2
			if m >= x {
				res = append(res, []int{x, m})
			}
		}
		// From y direction subtract
		if y >= x && y <= 2*x {
			res = append(res, []int{x, y - x})
		}
		return res
	}

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		x, y, steps := curr[0], curr[1], curr[2]

		if x == sx && y == sy {
			return steps
		}

		prevs := getPrevs(x, y)
		for _, p := range prevs {
			px, py := p[0], p[1]
			if px >= 0 && py >= 0 {
				k := keyFunc(px, py)
				if !visited[k] {
					visited[k] = true
					queue = append(queue, []int{px, py, steps + 1})
				}
			}
		}
	}

	return -1
}
# @lc code=end

File: 3609.minimum-moves-to-reach-target-in-grid.java
#
# @lc app=leetcode id=3609 lang=java
#
# [3609] Minimum Moves to Reach Target in Grid
#

# @lc code=start
import java.util.*;

class Solution {
    public int minMoves(int sx, int sy, int tx, int ty) {
        if (sx == tx && sy == ty) {
            return 0;
        }
        long SX = sx;
        long SY = sy;
        long TX = tx;
        long TY = ty;
        Set<Long> visited = new HashSet<>();
        Queue<long[]> queue = new LinkedList<>();
        long initKey = (TX << 32) | TY;
        visited.add(initKey);
        queue.offer(new long[]{TX, TY, 0L});
        while (!queue.isEmpty()) {
            long[] curr = queue.poll();
            long x = curr[0];
            long y = curr[1];
            long dist = curr[2];
            List<long[]> prevs = new ArrayList<>();
            // Undo add to x
            if (x >= y && x < 2L * y) {
                prevs.add(new long[]{x - y, y});
            }
            if (x % 2 == 0) {
                long half = x / 2;
                if (half >= y) {
                    prevs.add(new long[]{half, y});
                }
            }
            // Undo add to y
            if (y >= x && y < 2L * x) {
                prevs.add(new long[]{x, y - x});
            }
            if (y % 2 == 0) {
                long half = y / 2;
                if (half >= x) {
                    prevs.add(new long[]{x, half});
                }
            }
            for (long[] p : prevs) {
                long px = p[0];
                long py = p[1];
                long key = (px << 32) | py;
                if (visited.contains(key)) {
                    continue;
                }
                if (px == SX && py == SY) {
                    return (int)(dist + 1);
                }
                visited.add(key);
                queue.offer(new long[]{px, py, dist + 1});
            }
        }
        return -1;
    }
}
# @lc code=end